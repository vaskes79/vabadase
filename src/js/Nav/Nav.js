class Nav {
  constructor(
    nav = '.nav',
    activeModifire = 'nav--close',
    btn = '.nav__logo',
    linksContainer = '.nav__list'
  ) {
    this.nav = document.querySelector(nav);
    this.btn = document.querySelector(btn);
    this.links = document.querySelector(linksContainer);
    this.activeModifire = activeModifire;
  }

  addHandlers() {
    this.nav.classList.add(this.activeModifire);
    this.btn.addEventListener('click', this.clickHandler);
    this.links.addEventListener('click', this.clickHandler);
  }

  removeHandlers() {
    this.nav.classList.remove(this.activeModifire);
    this.btn.removeEventListener('click', this.clickHandler);
    this.links.removeEventListener('click', this.clickHandler);
  }

  resizeHandler = () => {
    window.addEventListener('resize', e => {
      let width = e.currentTarget.innerWidth;
      if (width >= 900) {
        this.removeHandlers();
      } else if (width < 900) {
        this.addHandlers();
      }
    });
  };

  clickHandler = () => {
    this.nav.classList.toggle(this.activeModifire);
  };

  init() {
    if (window.innerWidth < 900) {
      this.nav.classList.add(this.activeModifire);
      this.addHandlers();
    }

    this.resizeHandler();
  }
}

export default Nav;
