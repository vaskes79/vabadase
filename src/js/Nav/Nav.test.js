import Nav from './Nav';
let nav = null;
let mock = null;
let clickHandlerMock = null;

beforeEach(() => {
  nav = new Nav();
  nav.init();
  mock = jest.fn();
  clickHandlerMock = jest.spyOn(nav, 'clickHandler');
});

afterEach(() => {
  nav = null;
  mock = null;
  clickHandlerMock = null;
});

describe('Create instance and render', () => {
  it('expect that nav instance of Nav', () => {
    expect(nav instanceof Nav).toBeTruthy();
  });

  it('render components into DOM', () => {
    expect(nav).toMatchSnapshot();
  });
});

describe('props', () => {
  describe('default props', () => {
    it('this.nav exists has class .nav', () => {
      expect(nav.nav).not.toBeNil();
      expect(nav.nav.classList.contains('nav')).toBeTrue();
    });

    it('this.btn exists has class .nav__logo', () => {
      expect(nav.btn).not.toBeNil();
      expect(nav.btn.classList.contains('nav__logo')).toBeTrue();
    });

    it('this.links exists has class .nav__logo', () => {
      expect(nav.links).not.toBeNil();
      expect(nav.links.classList.contains('nav__list')).toBeTrue();
    });

    it('this.activeModifire exists has value nav--close', () => {
      expect(nav.activeModifire).not.toBeNil();
      expect(nav.activeModifire).toBeString();
      expect(nav.activeModifire).toBe('nav--close');
    });
  });
});

describe('methods', () => {
  const listMethods = [
    'addHandlers',
    'removeHandlers',
    'resizeHandler',
    'clickHandler',
    'init',
  ];

  listMethods.forEach(method => {
    describe(`${method}`, () => {
      it(`${method} exists`, () => {
        expect(nav[method]).toBeFunction();
      });

      it(`${method} toBeCalled`, () => {
        nav[method] = mock;
        nav[method]();
        expect(nav[method]).toBeCalled();
      });
    });
  });
});

describe('events', () => {
  it(`resize: innerWidth < 900`, () => {
    resizeTo(890);
    expect(nav.nav.classList.contains('nav--close')).toBeTrue();
  });

  it(`resize: innerWidth > 900`, () => {
    resizeTo(910);
    expect(nav.nav.classList.contains('nav--close')).toBeFalse();
  });

  it(`nav.btn: click`, () => {
    resizeTo(890);
    simulaitClick(nav.btn);
    expect(clickHandlerMock).toBeCalled();
    expect(clickHandlerMock).toHaveBeenCalledTimes(1);
  });

  it(`nav.links: click`, () => {
    resizeTo(890);
    simulaitClick(nav.links);
    expect(clickHandlerMock).toBeCalled();
    expect(clickHandlerMock).toHaveBeenCalledTimes(1);
  });
});
