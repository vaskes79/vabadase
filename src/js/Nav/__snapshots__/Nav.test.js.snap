// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create instance and render render components into DOM 1`] = `
Nav {
  "activeModifire": "nav--close",
  "btn": <div
    class="nav__logo"
  >
     
    <span
      class="logo"
    >
      vabaduse
    </span>
  </div>,
  "clickHandler": [MockFunction],
  "links": <ul
    class="nav__list"
  >
    <li
      class="nav__item"
    >
      <a
        class="nav__link"
        href="#home"
      >
        главная
      </a>
    </li>
    <li
      class="nav__item"
    >
      <a
        class="nav__link"
        href="#gallery"
      >
        эскизы
      </a>
    </li>
    <li
      class="nav__item"
    >
      <a
        class="nav__link"
        href="#price"
      >
        стоимость
      </a>
    </li>
    <li
      class="nav__item"
    >
      <a
        class="nav__link"
        href="#plans"
      >
        планировки
      </a>
    </li>
    <li
      class="nav__item"
    >
      <a
        class="nav__link"
        href="#times"
      >
        сроки
      </a>
    </li>
    <li
      class="nav__item"
    >
      <a
        class="nav__link"
        href="#contacts"
      >
        контакты
      </a>
    </li>
  </ul>,
  "nav": <nav
    class="nav"
  >
    <div
      class="nav__logo"
    >
       
      <span
        class="logo"
      >
        vabaduse
      </span>
    </div>
    <ul
      class="nav__list"
    >
      <li
        class="nav__item"
      >
        <a
          class="nav__link"
          href="#home"
        >
          главная
        </a>
      </li>
      <li
        class="nav__item"
      >
        <a
          class="nav__link"
          href="#gallery"
        >
          эскизы
        </a>
      </li>
      <li
        class="nav__item"
      >
        <a
          class="nav__link"
          href="#price"
        >
          стоимость
        </a>
      </li>
      <li
        class="nav__item"
      >
        <a
          class="nav__link"
          href="#plans"
        >
          планировки
        </a>
      </li>
      <li
        class="nav__item"
      >
        <a
          class="nav__link"
          href="#times"
        >
          сроки
        </a>
      </li>
      <li
        class="nav__item"
      >
        <a
          class="nav__link"
          href="#contacts"
        >
          контакты
        </a>
      </li>
    </ul>
    <div
      class="nav__phone phone"
    >
      +7(921) 902-51-13
    </div>
  </nav>,
  "resizeHandler": [Function],
}
`;
